server/src/main/java/org/elasticsearch/search/sort/SortValue.java

```diff
@@ -128,6 +125,9 @@ public final XContentBuilder toXContent(XContentBuilder builder, DocValueFormat
     @Override
     public abstract String toString();
 
+    // Force implementations to override compareToDifferentType and sort each subclass by type first
+    protected abstract int compareToDifferentType();
```

Comment:
maybe `typeComparisonKey` is a better name.

Reply:
This name was meant to be the counterpart of `compareToSameType`. I thought it just reads nice like this:
```
private static final Comparator<SortValue> TYPE_COMPARATOR = Comparator.comparing(SortValue::compareToDifferentType);
    private static final Comparator<SortValue> VALUE_COMPARATOR = TYPE_COMPARATOR.thenComparing(SortValue::compareToSameType);
```

you can read it almost as "compare to different type and then compare to same type"